### PR TITLE
Bug 1903206: Add unit tests to verify NotReadyAddresses in EndpointSlices

### PR DESCRIPTION
**Pull request by frobware**
Moved the existing utility function into endpointsubset package. Added
unit tests to assert on EndpointSlices having Ready/NotReady
conditions.

Follow on from https://github.com/openshift/router/pull/229

### CODE REVIEW

File: pkg/router/controller/endpointsubset/converter.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package endpointsubset
+
+import (
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/discovery/v1beta1"
+)
+
+// ConvertEndpointSlice converts items to a slice of EndpointSubset's.
+func ConvertEndpointSlice(items []v1beta1.EndpointSlice, addressOrderByFuncs []EndpointAddressLessFunc, portOrderByFuncs []EndpointPortLessFunc) []v1.EndpointSubset {
```
Comment (candita) [inline pkg/router/controller/endpointsubset/converter.go:9]: Perhaps if you always use the same "order by" function each time, you can remove it from the arguments and hard code it in the function.  Would make cleaner tests.

File: pkg/router/controller/endpointsubset/converter_test.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
+package endpointsubset_test
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/api/core/v1"
+	discoveryv1beta1 "k8s.io/api/discovery/v1beta1"
```
Comment (candita) [inline pkg/router/controller/endpointsubset/converter_test.go:8]: It wouild be helpful if you could use the same import aliases in the test as in the tested code.

Comment (frobware) [reply pkg/router/controller/endpointsubset/converter_test.go]: Done in https://github.com/openshift/router/commit/26cbd10beee116fcf9f8054265a0e72eb213b5e1.

File: pkg/router/controller/endpointsubset/converter.go (hunk 1 above)
Comment (frobware) [reply pkg/router/controller/endpointsubset/converter.go:9]: Yep, good call. done in https://github.com/openshift/router/commit/26cbd10beee116fcf9f8054265a0e72eb213b5e1

Comment (sgreene570) [reply pkg/router/controller/endpointsubset/converter.go:9]: > Yep, good call. done in [26cbd10](https://github.com/openshift/router/commit/26cbd10beee116fcf9f8054265a0e72eb213b5e1)

Did these changes make it into `26cbd10`? I think I might still be seeing the original code?

Comment (frobware) [reply pkg/router/controller/endpointsubset/converter.go:9]: https://github.com/openshift/router/pull/231/commits/26cbd10beee116fcf9f8054265a0e72eb213b5e1#diff-eb4119cc4334b78b1a123bdf8b86d9898af5909d18b0becc0daac1c000f8ac81R100

Comment (sgreene570) [reply pkg/router/controller/endpointsubset/converter.go:9]: Ah I see, a middle ground was reached here. `ConvertEndpointSlice` still has the `addressOrderByFuncs` and `portOrderByFuncs` parameters, which is good in the event that we want to use different ordering funcs in the future. But the test cases however, don't specify specific `<foo>OrderByFuncs` and instead get tested with the defaults. Makes sense to me.

**Comment by frobware** [comment]
test failures:

```
failed to acquire lease: resources not found
```

/retest

File: pkg/router/controller/endpointsubset/converter.go (hunk 1 above)
Comment (candita) [reply pkg/router/controller/endpointsubset/converter.go:9]: That's fine by me.

**Comment by frobware** [comment]
> ```
> package endpointsubset
> ```

Although this has merged I disagree. In general always test the public API, using _test was a deliberate choice.

**Comment by frobware** [comment]
> @frobware: #231 failed to apply on top of branch "release-4.6":
> 
> ```
> Applying: Add unit tests to verify NotReadyAddresses in EndpointSlices
> Using index info to reconstruct a base tree...
> M	pkg/router/controller/router_controller.go
> Falling back to patching base and 3-way merge...
> Auto-merging pkg/router/controller/router_controller.go
> CONFLICT (content): Merge conflict in pkg/router/controller/router_controller.go
> error: Failed to merge in the changes.
> hint: Use 'git am --show-current-patch=diff' to see the failed patch
> Patch failed at 0001 Add unit tests to verify NotReadyAddresses in EndpointSlices
> When you have resolved this problem, run "git am --continue".
> If you prefer to skip this patch, run "git am --skip" instead.
> To restore the original branch and stop patching, run "git am --abort".
> ```

The merge conflict will clear up when https://github.com/openshift/router/pull/230 merges.

**Comment by frobware** [comment]
> ```
> 				t.Errorf("ConvertEndpointSlice() failed (-got +want):\n%s", diff)
> ```



> @frobware: #231 failed to apply on top of branch "release-4.6":
> 
> ```
> Applying: Add unit tests to verify NotReadyAddresses in EndpointSlices
> Using index info to reconstruct a base tree...
> M	pkg/router/controller/router_controller.go
> Falling back to patching base and 3-way merge...
> Auto-merging pkg/router/controller/router_controller.go
> CONFLICT (content): Merge conflict in pkg/router/controller/router_controller.go
> error: Failed to merge in the changes.
> hint: Use 'git am --show-current-patch=diff' to see the failed patch
> Patch failed at 0001 Add unit tests to verify NotReadyAddresses in EndpointSlices
> When you have resolved this problem, run "git am --continue".
> If you prefer to skip this patch, run "git am --skip" instead.
> To restore the original branch and stop patching, run "git am --abort".
> ```

The cherry-pick needs https://github.com/openshift/router/pull/230

**Comment by Miciah** [comment]
> > ```
> > package endpointsubset
> > ```
> 
> Although this has merged I disagree. In general always test the public API, using _test was a deliberate choice.

Interesting.  I didn't realize that that was allowed, but
```console
% go help test
usage: go test [build/test flags] [packages] [build/test flags & test binary flags]

[...]

Test files that declare a package with the suffix "_test" will be compiled as a
separate package, and then linked and run with the main test binary.
```
https://stackoverflow.com/questions/19998250/proper-package-naming-for-testing-with-the-go-language

TIL.

**Pull request merged**